src/cursor/cur_hs.c

```diff
@@ -381,9 +381,17 @@ __curhs_prev_visible(WT_SESSION_IMPL *session, WT_CURSOR_HS *hs_cursor)
 
         /*
          * If the stop time pair on the tombstone in the history store is already globally visible
-         * we can skip it.
+         * and the key is set, we should stop here as none of the key versions before the current
+         * version of the key will be visible any way. Only exception is when history store contains
+         * a deleted update without a timestamp.
          */
         if (__wt_txn_tw_stop_visible_all(session, &cbt->upd_value->tw)) {
```

Comment:
I think this change may not be correct. This is to skip globally visible records. There may be records with non-globally visible tombstones before that.

Reply:
The following part of the code should help us stop early:

```
      if (__wt_txn_tw_stop_visible(session, &cbt->upd_value->tw)) {
            /*
             * If the stop time point of a record is visible to us, we won't be able to see anything
             * for this entire key.
             */
            if (F_ISSET(hs_cursor, WT_HS_CUR_KEY_SET)) {
                ret = WT_NOTFOUND;
                goto err;
            } else
                continue;
        }
```